### ISSUE TITLE
Remove bam_to_bqsr merge inputs after the merge.

### ISSUE THREAD

**Issue by jasonwalker80** [genome/analysis-workflows#621]
Once the merge step is complete. The inputs, per instrument data aligned BAMs, should no longer be necessary. We could remove those inputs at this point in the alignment workflow.

https://github.com/genome/analysis-workflows/blob/e509210450e4c62aecb99a228fc97f0eae2d9580/definitions/subworkflows/bam_to_bqsr.cwl#L58

**Comment by jasonwalker80** [comment]
Summary of all intermediate BAMs that can be removed:
After the merge, the align/tagged_bam inputs could be removed.
After the name_sort, the merge/merged_bam input could be removed.
After mark_duplicates_and_sort, the name_sort/name_sorted_bam can be removed.
After apply_bqsr, the mark_duplicates_and_sort/sorted_bam can be removed.

**Comment by jasonwalker80** [comment]
I can confirm that BAM files are no longer created many times for downstream steps. However, the alignment step itself, creating intermediate copies of the BAMs, is still using a lot of unnecessary space. For example:
```
792G    /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-alignment_and_qc
30G     /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-bam_to_cram
21G     /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-detect_variants
30G     /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-index_cram
12G     /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-variant_callers
882G    total
```

Within the alignment step the breakdown of storage usage looks like:
```
256G    /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-alignment_and_qc/wgs_alignment.cwl/d7086dba-3c19-434a-9884-7a9c55d5c809/call-alignment/bam_to_bqsr.cwl/265110be-6c72-4b24-bd86-d218d525fdc6/call-align
155G    /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-alignment_and_qc/wgs_alignment.cwl/d7086dba-3c19-434a-9884-7a9c55d5c809/call-alignment/bam_to_bqsr.cwl/265110be-6c72-4b24-bd86-d218d525fdc6/call-apply_bqsr
1.6G    /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-alignment_and_qc/wgs_alignment.cwl/d7086dba-3c19-434a-9884-7a9c55d5c809/call-alignment/bam_to_bqsr.cwl/265110be-6c72-4b24-bd86-d218d525fdc6/call-bqsr
75G     /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-alignment_and_qc/wgs_alignment.cwl/d7086dba-3c19-434a-9884-7a9c55d5c809/call-alignment/bam_to_bqsr.cwl/265110be-6c72-4b24-bd86-d218d525fdc6/call-index_bam
153G    /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-alignment_and_qc/wgs_alignment.cwl/d7086dba-3c19-434a-9884-7a9c55d5c809/call-alignment/bam_to_bqsr.cwl/265110be-6c72-4b24-bd86-d218d525fdc6/call-mark_duplicates_and_sort
153G    /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-alignment_and_qc/wgs_alignment.cwl/d7086dba-3c19-434a-9884-7a9c55d5c809/call-alignment/bam_to_bqsr.cwl/265110be-6c72-4b24-bd86-d218d525fdc6/call-merge
256K    /gscmnt/gc2648/wgs/model_data/aa00989be76243d6bc544ce891335da6/build6bf88831c26145849cca97635c3e967e/tmp/cromwell-executions/germline_wgs.cwl/adf00e15-8017-43a7-8455-0a8de0a48697/call-alignment_and_qc/wgs_alignment.cwl/d7086dba-3c19-434a-9884-7a9c55d5c809/call-alignment/bam_to_bqsr.cwl/265110be-6c72-4b24-bd86-d218d525fdc6/call-name_sort
791G    total
```

**Comment by chrisamiller** [comment]
Are there thoughts on whether this is possible within the CWL workflow? Does it have to do something janky like call a script that traverses the workflow tree and finds the right bams to remove, or is there some CWL function that allows for "cleanup"?

**Comment by tmooney** [comment]
I'm sure there's some way to do it, but it would be outside the intended bounds of CWL.  How to do it would be backend and implementation dependent--that is, different for `cwltool` than for Cromwell and different depending on your environment (local filesystem, GCP, etc.).  This would be better implemented as a feature of the workflow runner.  (The GMS cleans up intermediate files at the end, for instance, and Cromwell has an experimental `delete_intermediate_output_files` option for GCP.  But both operate at the end of workflows--there's an assumption you have enough working space to hold all the intermediate files.)

For the case of these alignment files, moving to a speedseq-ish giant single step that does the alignment would reduce the duplication.

**Comment by chrisamiller** [comment]
okay, given that limitation, I'll suggest that we close this issue.  Thoughts on whether we should create a new issue that references this and raises the possibility of a speedseq-like integrated alignment step (outside the bounds of our 2.0 milestone)?
